File: Habitica/src/main/java/com/habitrpg/android/habitica/helpers/notifications/PushNotificationManager.java
package com.habitrpg.android.habitica.helpers.notifications;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.firebase.iid.FirebaseInstanceId;
import com.google.firebase.messaging.RemoteMessage;
import com.habitrpg.android.habitica.HabiticaBaseApplication;
import com.habitrpg.android.habitica.data.ApiClient;
import com.habitrpg.android.habitica.models.HabitRPGUser;
import com.habitrpg.android.habitica.models.PushDevice;

import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;

public class PushNotificationManager {

    static final String PARTY_INVITE_PUSH_NOTIFICATION_KEY = "invitedParty";
    static final String RECEIVED_PRIVATE_MESSAGE_PUSH_NOTIFICATION_KEY = "newPM";
    static final String RECEIVED_GEMS_PUSH_NOTIFICATION_KEY = "giftedGems";
    static final String RECEIVED_SUBSCRIPTION_GIFT_PUSH_NOTIFICATION_KEY = "giftedSubscription";
    static final String GUILD_INVITE_PUSH_NOTIFICATION_KEY = "invitedGuild";
    static final String QUEST_INVITE_PUSH_NOTIFICATION_KEY = "questInvitation";
    static final String QUEST_BEGUN_PUSH_NOTIFICATION_KEY = "questStarted";
    static final String WON_CHALLENGE_PUSH_NOTIFICATION_KEY = "wonChallenge";
    private static final String DEVICE_TOKEN_PREFERENCE_KEY = "device-token-preference";
    private static PushNotificationManager instance = null;
    @Inject
    public ApiClient apiClient;

    private String refreshedToken;
    private SharedPreferences sharedPreferences;
    private Context context;
    private HabitRPGUser user;

    protected PushNotificationManager(Context context) {
        HabiticaBaseApplication.getComponent().inject(this);
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
    }

    public static PushNotificationManager getInstance(Context context) {
        if (instance == null) {
            instance = new PushNotificationManager(context.getApplicationContext());
        }

        instance.refreshedToken = instance.sharedPreferences.getString(DEVICE_TOKEN_PREFERENCE_KEY, "");
        instance.context = context.getApplicationContext();

        return instance;
    }

    public void setUser(HabitRPGUser user) {
        this.user = user;
    }

    void setRefreshedToken(String refreshedToken) {
        if (this.refreshedToken == null) {
            return;
        }

        this.refreshedToken = refreshedToken;
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(DEVICE_TOKEN_PREFERENCE_KEY, refreshedToken);
        editor.apply();
    }

    //@TODO: Use preferences
    public void addPushDeviceUsingStoredToken() {
        if (this.refreshedToken == null || this.refreshedToken.isEmpty()) {
            this.refreshedToken = FirebaseInstanceId.getInstance().getToken();
        }

        if (this.refreshedToken == null || this.refreshedToken.isEmpty()) {
            return;
        }

        if (this.user == null || this.userHasPushDevice()) {
            return;
        }

        if (!this.userIsSubscribedToNotifications()) {
            return;
        }

        Map<String, String> pushDeviceData = new HashMap<>();
        pushDeviceData.put("regId", this.refreshedToken);
        pushDeviceData.put("type", "android");
        apiClient.addPushDevice(pushDeviceData)

            .subscribe(aVoid -> {}, throwable -> {});
    }

    public void removePushDeviceUsingStoredToken() {
        apiClient.deletePushDevice(this.refreshedToken)

            .subscribe(aVoid -> {}, throwable -> {});
    }

    private Boolean userHasPushDevice() {
        if (this.user.getPushDevices() == null) {
            return true;
        }

        for (PushDevice pushDevice : this.user.getPushDevices()) {
            if (pushDevice.getRegId().equals(this.refreshedToken)) {
                return true;
            }
        }
        return false;
    }

    void displayNotification(RemoteMessage remoteMessage) {
        String remoteMessageIdentifier = remoteMessage.getData().get("identifier");

        HabiticaLocalNotificationFactory notificationFactory = new HabiticaLocalNotificationFactory();
        HabiticaLocalNotification notification = notificationFactory.build(remoteMessageIdentifier);
        if (userIsSubscribedToNotificationType(remoteMessageIdentifier) && notification != null) {
            notification.setExtras(remoteMessage.getData());
            notification.notifyLocally(this.context, remoteMessage.getData().get("title"), remoteMessage.getData().get("body"));
        }
    }

    private boolean userIsSubscribedToNotifications() {
        return sharedPreferences.getBoolean("pushNotifications", true);
    }

    private boolean userIsSubscribedToNotificationType(String type) {
        String key = "";

        //@TODO: If user has push turned off to send

        if (type.equals(PARTY_INVITE_PUSH_NOTIFICATION_KEY)) {
            key = "preference_push_invited_to_party";
        } else if (type.contains(RECEIVED_PRIVATE_MESSAGE_PUSH_NOTIFICATION_KEY)) {
            key = "preference_push_received_a_private_message";
        } else if (type.contains(RECEIVED_GEMS_PUSH_NOTIFICATION_KEY)) {
            key = "preference_push_gifted_gems";
        } else if (type.contains(RECEIVED_SUBSCRIPTION_GIFT_PUSH_NOTIFICATION_KEY)) {
            key = "preference_push_gifted_subscription";
        } else if (type.contains(GUILD_INVITE_PUSH_NOTIFICATION_KEY)) {
            key = "preference_push_invited_to_guild";
        } else if (type.contains(QUEST_INVITE_PUSH_NOTIFICATION_KEY)) {
            key = "preference_push_invited_to_quest";
        } else if (type.contains(QUEST_BEGUN_PUSH_NOTIFICATION_KEY)) {
            key = "preference_push_your_quest_has_begun";
        } else if (type.contains(WON_CHALLENGE_PUSH_NOTIFICATION_KEY)) {
            key = "preference_push_you_won_challenge";
        }

        return sharedPreferences.getBoolean(key, true);
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/helpers/notifications/PushNotificationManager.kt
package com.habitrpg.android.habitica.helpers.notifications

import android.content.Context
import android.content.SharedPreferences
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.edit
import com.google.firebase.messaging.FirebaseMessaging
import com.google.firebase.messaging.RemoteMessage
import com.habitrpg.android.habitica.data.ApiClient
import com.habitrpg.android.habitica.helpers.Analytics
import com.habitrpg.android.habitica.helpers.EventCategory
import com.habitrpg.android.habitica.helpers.HitType
import com.habitrpg.android.habitica.models.user.User
import com.habitrpg.common.habitica.helpers.launchCatching
import kotlinx.coroutines.MainScope
import java.io.IOException

class PushNotificationManager(
    var apiClient: ApiClient,
    private val sharedPreferences: SharedPreferences,
    private val context: Context
) {

    var refreshedToken: String = ""
        set(value) {
            if (value.isEmpty()) {
                return
            }

            field = value
            sharedPreferences.edit {
                putString(DEVICE_TOKEN_PREFERENCE_KEY, value)
            }
        }
    private var user: User? = null

    fun setUser(user: User) {
        this.user = user
    }

    /**
     * New installs on Android 13 require
     * Notification permissions be approved.
     * Devices on Android 12L or lower with previously
     * allowed notification permissions that update to 13
     * will have notification permissions enabled by default.
     */
    fun notificationPermissionEnabled(): Boolean {
        val notificationManager = NotificationManagerCompat.from(context)
        return notificationManager.areNotificationsEnabled()
    }

    fun addPushDeviceUsingStoredToken() {
        if (refreshedToken.isNotBlank()) {
            addRefreshToken()
        } else {
            try {
                FirebaseMessaging.getInstance().token.addOnCompleteListener {
                    try {
                        refreshedToken = it.result
                        addRefreshToken()
                    } catch (_: IOException) {
                        // catchy catch-catch
                    } catch (_: Exception) {
                        // catchy catch-catch-cat, I'm out of breath.
                    }
                }
            } catch (_: Exception) {
                // catchy catch
            }
        }
    }

    private fun addRefreshToken() {
        if (this.refreshedToken.isEmpty() || this.user == null || this.userHasPushDevice()) {
            return
        }
        val pushDeviceData = HashMap<String, String>()
        pushDeviceData["regId"] = this.refreshedToken
        pushDeviceData["type"] = "android"
        MainScope().launchCatching {
            apiClient.addPushDevice(pushDeviceData)
        }
    }

    suspend fun removePushDeviceUsingStoredToken() {
        if (this.refreshedToken.isEmpty() || !userHasPushDevice()) {
            return
        }
        apiClient.deletePushDevice(refreshedToken)
    }

    private fun userHasPushDevice(): Boolean {
        for (pushDevice in this.user?.pushDevices ?: emptyList()) {
            if (pushDevice.regId == this.refreshedToken) {
                return true
            }
        }
        return this.user?.pushDevices == null
    }

    private fun userIsSubscribedToNotificationType(type: String?): Boolean {
        val key = when {
            type == PARTY_INVITE_PUSH_NOTIFICATION_KEY -> "preference_push_invited_to_party"
            type?.contains(RECEIVED_PRIVATE_MESSAGE_PUSH_NOTIFICATION_KEY) == true -> "preference_push_received_a_private_message"
            type?.contains(RECEIVED_GEMS_PUSH_NOTIFICATION_KEY) == true -> "preference_push_gifted_gems"
            type?.contains(RECEIVED_SUBSCRIPTION_GIFT_PUSH_NOTIFICATION_KEY) == true -> "preference_push_gifted_subscription"
            type?.contains(GUILD_INVITE_PUSH_NOTIFICATION_KEY) == true -> "preference_push_invited_to_guild"
            type?.contains(QUEST_INVITE_PUSH_NOTIFICATION_KEY) == true -> "preference_push_invited_to_quest"
            type?.contains(QUEST_BEGUN_PUSH_NOTIFICATION_KEY) == true -> "preference_push_your_quest_has_begun"
            type?.contains(WON_CHALLENGE_PUSH_NOTIFICATION_KEY) == true -> "preference_push_you_won_challenge"
            type?.contains(CONTENT_RELEASE_NOTIFICATION_KEY) == true -> "preference_push_content_release"
            else -> return true
        }

        return sharedPreferences.getBoolean(key, true)
    }

    companion object {
        const val PARTY_INVITE_PUSH_NOTIFICATION_KEY = "invitedParty"
        const val RECEIVED_PRIVATE_MESSAGE_PUSH_NOTIFICATION_KEY = "newPM"
        const val RECEIVED_GEMS_PUSH_NOTIFICATION_KEY = "giftedGems"
        const val RECEIVED_SUBSCRIPTION_GIFT_PUSH_NOTIFICATION_KEY = "giftedSubscription"
        const val GUILD_INVITE_PUSH_NOTIFICATION_KEY = "invitedGuild"
        const val QUEST_INVITE_PUSH_NOTIFICATION_KEY = "questInvitation"
        const val QUEST_BEGUN_PUSH_NOTIFICATION_KEY = "questStarted"
        const val WON_CHALLENGE_PUSH_NOTIFICATION_KEY = "wonChallenge"
        const val CHANGE_USERNAME_PUSH_NOTIFICATION_KEY = "changeUsername"
        const val GIFT_ONE_GET_ONE_PUSH_NOTIFICATION_KEY = "gift1get1"
        const val CHAT_MENTION_NOTIFICATION_KEY = "chatMention"
        const val GROUP_ACTIVITY_NOTIFICATION_KEY = "groupActivity"
        const val CONTENT_RELEASE_NOTIFICATION_KEY = "contentRelease"
        const val G1G1_PROMO_KEY = "g1g1Promo"
        private const val DEVICE_TOKEN_PREFERENCE_KEY = "device-token-preference"

        fun displayNotification(remoteMessage: RemoteMessage, context: Context, pushNotificationManager: PushNotificationManager? = null) {
            val remoteMessageIdentifier = remoteMessage.data["identifier"]

            if (pushNotificationManager?.userIsSubscribedToNotificationType(remoteMessageIdentifier) != false) {
                if (remoteMessage.data.containsKey("sendAnalytics")) {
                    val additionalData = HashMap<String, Any>()
                    additionalData["identifier"] = remoteMessageIdentifier ?: ""
                    Analytics.sendEvent(
                        "receive notification",
                        EventCategory.BEHAVIOUR,
                        HitType.EVENT,
                        additionalData
                    )
                }

                val notificationFactory = HabiticaLocalNotificationFactory()
                val localNotification = notificationFactory.build(
                    remoteMessageIdentifier,
                    context
                )
                localNotification.setExtras(remoteMessage.data)
                val notification = remoteMessage.notification
                if (notification != null) {
                    localNotification.notifyLocally(
                        notification.title ?: remoteMessage.data["title"],
                        notification.body ?: remoteMessage.data["body"],
                        remoteMessage.data
                    )
                } else {
                    localNotification.notifyLocally(
                        remoteMessage.data["title"],
                        remoteMessage.data["body"],
                        remoteMessage.data
                    )
                }
            }
        }
    }
}
